server/nodeset/upload-deposit-data.go

```diff
@@ -68,79 +69,109 @@ func (c *nodesetUploadDepositDataContext) PrepareData(data *swapi.NodesetUploadD
 	}
-	data.TotalCount = uint64(len(keys))
 
-	// Find the ones that haven't been uploaded yet
-	unregisteredKeys := []*eth2types.BLSPrivateKey{}
-	newPubkeys := []beacon.ValidatorPubkey{}
-	for _, key := range keys {
-		pubkey := beacon.ValidatorPubkey(key.PublicKey().Marshal())
-		_, exists := registeredPubkeyMap[pubkey]
-		if !exists {
-			unregisteredKeys = append(unregisteredKeys, key)
-			newPubkeys = append(newPubkeys, pubkey)
+	activePubkeysOnNodeset := []beacon.ValidatorPubkey{}
+	pendingPubkeysOnNodeset := []beacon.ValidatorPubkey{}
+	for _, validator := range nodesetStatusResponse {
+		_, exists := publicKeyMap[validator.Pubkey]
+		if exists {
+			publicKeyMap[validator.Pubkey] = false
+
+			if validator.Status != pendingState {
+				activePubkeysOnNodeset = append(activePubkeysOnNodeset, validator.Pubkey)
+			} else {
+				pendingPubkeysOnNodeset = append(pendingPubkeysOnNodeset, validator.Pubkey)
+			}
 		}
 	}
-	data.UnregisteredPubkeys = newPubkeys
 
-	if len(unregisteredKeys) == 0 {
-		return types.ResponseStatus_Success, nil
+	newPublicKeys := []beacon.ValidatorPubkey{}
+	for pubkey, unprocessed := range publicKeyMap {
+		if unprocessed {
+			newPublicKeys = append(newPublicKeys, pubkey)
+		}
 	}
+	publicKeys = newPublicKeys
 
-	// Make sure validator has enough funds to pay for the deposit
-	if !c.bypassBalanceCheck {
-		balance, err := ec.BalanceAt(ctx, opts.From, nil)
-		if err != nil {
-			return types.ResponseStatus_Error, fmt.Errorf("error getting balance: %w", err)
+	// Process public keys based on their status
+	unregisteredKeys := []*eth2types.BLSPrivateKey{}
+	data.TotalCount = uint64(len(publicKeys))
+	// Used for displaying the unregistered keys in the response
+	unregisteredPubkeys := []beacon.ValidatorPubkey{}
+
+	for i, pubkey := range publicKeys {
+		if !swcommon.IsUploadedToNodeset(pubkey, activePubkeysOnNodeset) {
+			unregisteredKeys = append(unregisteredKeys, privateKeys[i])
+			unregisteredPubkeys = append(unregisteredPubkeys, pubkey)
 		}
-		data.Balance = balance
+	}
 
-		totalCost := big.NewInt(int64(len(unregisteredKeys)))
-		totalCost.Mul(totalCost, eth.EthToWei(0.01))
-		data.RequiredBalance = totalCost
+	// Determine if sufficient balance is available for deposits
+	balance, err := ec.BalanceAt(ctx, opts.From, nil)
+	if err != nil {
+		return types.ResponseStatus_Error, fmt.Errorf("error getting balance: %w", err)
+	}
+	data.Balance = balance
+
+	totalCost := new(big.Int)
+
+	// Add deposit cost for each unregistered key
+	for range unregisteredKeys {
+		totalCost.Add(totalCost, eth.EthToWei(validatorDepositCost))
+	}
+	// Total cost needs to account for pending validators
+	for range pendingPubkeysOnNodeset {
+		totalCost = totalCost.Add(totalCost, eth.EthToWei(validatorDepositCost))
+	}
+
+	data.SufficientBalance = (totalCost.Cmp(balance) <= 0)
+
+	if !data.SufficientBalance {
+		for len(unregisteredKeys) > 0 && totalCost.Cmp(balance) > 0 {
```

Comment:
I'm sure you can do this in-place during the loop above instead of adding everything and selectively subtracting things afterwards.

Reply:
resolved